python/whylogs/api/logger/multi_dataset_logger/multi_dataset_rolling_logger.py

```diff
@@ -0,0 +1,362 @@
+        elif isinstance(message, StatusMessage):
+            self._process_status_message(message)
+        else:
+            # Safe guard for forgetting to handle a message in development
+            raise Exception(f"Don't know how to handle message {message}")
+
+    def _process_status_message(self, message: StatusMessage) -> None:
+        profiles = 0
+        segment_caches = 0
+        for ts, container in self._cache.items():
+            if container._has_segments():
+                segment_caches += 1
+            else:
+                profiles += 1
+
+        writers = 0
+        writables = 0
+        for writer, stuff in self._writers.items():
+            writers += 1
+            writables += len(stuff)
+
+        status = LoggerStatus(
+            dataset_timestamps=len(self._cache),
+            dataset_profiles=profiles,
+            segment_caches=segment_caches,
+            writers=writers,
+            pending_writables=writables,
+        )
+        message.result.set_result(status)
+
+    def _process_close_message(self, message: CloseMessage) -> None:
+        self._timer.stop()
+        # Force wait for all writers to handle their pending items
+        self._process_flush_message(FlushMessage())
+        while self._has_pending():
+            self._process_flush_message(FlushMessage())
+
+    def _has_pending(self) -> bool:
+        has_pending = False
+        for writer, pending in self._writers.items():
+            has_pending = len(pending) > 0
+        return has_pending
+
+    def _process_flush_message(self, message: FlushMessage) -> None:
+        for dataset_timestamp, container in self._cache.items():
+            self._logger.debug(f"Generating result set for dataset timestamp {dataset_timestamp}")
+
+            for writer, pending in self._writers.items():
+                for writable in container.to_result_set().get_writables() or []:
+                    pending.append(PendingWritable(attempts=0, writable=writable))
+
+        self._cache = {}
+        self._write_pending()
+
+    def _write_pending(self) -> None:
+        new_state: Dict[Writer, List[PendingWritable]] = {}
+        for writer, pending in self._writers.items():
+            failures: List[PendingWritable] = []
+            self._logger.info(f"Writing out result set with {type(writer).__name__}")
+            for p in pending:
+                self._logger.debug(f"Writing {p.attempts} attempt")
+                failed = False
+                try:
+                    success, msg = writer.write(p.writable)
+
+                    if not success:
+                        self._logger.error(f"Couldn't write profile: {msg}")
+                        failed = True
+                except Exception as e:
+                    self._logger.exception(e)
+                    failed = True
+
+                if failed:
+                    p.attempts += 1
+                    if p.attempts < 3:
```

Comment:
consider config value for retry attempts?

Reply:
I think we just put this off for now until we have a reason to set it. It's easy to create that config later and I think three will probably be fine for a long long time.